### ISSUE TITLE
Why use internal stylesheets over external CSS files?

### ISSUE THREAD

**Issue by Hrishikesh-K** [gatsbyjs/gatsby#25809]
## Summary

Why does Gatsby use internal stylesheet as opposed to an external one for the entire website?

## Relevant information

I have worked on 4 to 5 Gatsby websites now (all were modified versions of different starters), but, there's one thing common in the output of all of these:

The CSS styling is included in the `<style>` tag in the `<head>` of each HTML document. The `<style>` tag does have a `data-href` attribute, but, I've no idea what it does.

All the websites I've made till now are blogs, so, basically, I use the same styles for all my blog pages. But, each page has the same styles hardcoded in the output HTML page.

So, I've been meaning to ask this question from long now, but, I ignored it. However, today, I noticed in Bing Search Console, I've got a warning on few of my pages stating: `Page size is over 125 KB and there are high risks of page not being fully cached while showing in the search engine results page (SERP)`. It's really not something I care about for now, but, it convinced me to ask this question here. According to my knowledge, using a single (or a few) external CSS files for the entire website would be faster, as the browser will have to download it once and for all other pages, the cached CSS file would work, instead of injecting same styles in all HTML pages. So, if someone can explain this behaviour to me, that'd help.

### Environment (if relevant)

```
System:
  OS: Windows 10 10.0.19041
  CPU: (8) x64 Intel(R) Core(TM) i5-8300H CPU @ 2.30GHz
Binaries:
  Node: 12.17.0 - C:\Program Files\nodejs\node.EXE
  npm: 6.14.6 - C:\Program Files\nodejs\npm.CMD
Languages:
  Python: 3.8.3
Browsers:
  Edge: 44.19041.1.0
npmGlobalPackages:
  gatsby-cli: 2.12.61
```
### File contents (if changed)

These are the 2 starters I've majorly used:

1. https://www.gatsbyjs.org/starters/panr/gatsby-starter-hello-friend/
1. https://www.gatsbyjs.org/starters/akzhy/gatsby-starter-elemental/

However, about the second one in the list, I have customised it to a level that I use a single style sheet now and inject it in `<head>` using `Helmet`. So, that's working fine too, adding to my curiosity. Here's a demo to that: https://hritech.netlify.app/. In this, Gatsby has still managed to inject some minimal styles:

```HTML
<style data-href = "/styles.4b3ce496e3b88ee7f750.css">
  .tl-edges
    {
      max-width: 100%;
      overflow-x: hidden;
    }
  .tl-wrapper
    {
      width: 100%;
      float: left;
      position: relative;
    }
  .tl-wrapper+.tl-wrapper
    {
      margin-left: -100%;
      margin-right: 0;
    }
</style>
```

I've no idea from where those styles are coming, but, they don't hurt, so, I don't bother.

Basically, some explanation about this behaviour would be great.

**Comment by pieh** [comment]
> It's really not something I care about for now, but, it convinced me to ask this question here. According to my knowledge, using a single (or a few) external CSS files for the entire website would be faster, as the browser will have to download it once and for all other pages, the cached CSS file would work, instead of injecting same styles in all HTML pages

Gatsby will load .html content once (first page you navigate to), it will not read those again on subsequent navigation (that happens purely client side). Because of that splitting it to separate css file just doesn't have those usual benefits as Gatsby use custom navigation, but still would have negative of making another network request for critical resource.

AMP uses similar approach where they even disallow external stylesheets (save from font ones) - https://amp.dev/documentation/guides-and-tutorials/develop/style_and_layout/style_pages/

All of this is to be able to have hydrated site painted in browser as quickly as possible

**Comment by Hrishikesh-K** [comment]
Oh, that's interesting, so, basically, the HTML for subsequent pages is never requested? Because as long as it's requesting it, the browser will download it resulting in extra bytes, right? Thank you for taking time to answer,

**Comment by herecydev** [comment]
> the HTML for subsequent pages is never requested

That's correct, instead it'll fetch page-data.json for each page

**Comment by Hrishikesh-K** [comment]
Ahh, now I know why the `page-data.json` exists. Thanks for the clarification guys. The more I am getting to know Gatsby, the more I seem to like it.
